scripts/test_env.py

```diff
@@ -0,0 +1,240 @@
+        os.environ["IROHA_PRIVATE_KEY"] = self.private_key
+        os.environ["SUMERAGI_DEBUG_FORCE_SOFT_FORK"] = "false"
+        os.environ["TORII_P2P_ADDR"] = f"{self.host_addr}:{self.p2p_port}"
+        os.environ["TORII_API_URL"] = f"{self.host_addr}:{self.api_port}"
+        os.environ["TORII_TELEMETRY_URL"] = f"{self.host_addr}:{self.telemetry_port}"
+        os.environ["TOKIO_CONSOLE_ADDR"] = f"{self.host_addr}:{self.tokio_console_port}"
+
+        genesis_arg = "--submit-genesis" if is_genesis else ""
+        # FD never gets closed
+        log_file = open(peer_dir.joinpath(".log"), "w")
+        # These processes are created detached from the parent process already
+        subprocess.Popen([iroha_bin, f"{genesis_arg}"],
+                    stdout=log_file, stderr=subprocess.STDOUT)
+
+def pos_int(arg):
+    if int(arg) > 0:
+        return int(arg)
+    else:
+        raise argparse.ArgumentTypeError(f"Argument {arg} must be a positive integer")
+
+def prompt_build_bin(bin_name: str, root_dir: pathlib.Path, target_dir: pathlib.Path):
+    logging.error(f"The binary `{bin_name}` wasn't found in `{root_dir}` directory")
+    while True:
+        prompt = input(f"Build it by running `cargo build --bin {bin_name}`? (y/n)\n")
+        if prompt.lower() in ["y", "yes"]:
+            subprocess.run(
+                ["cargo", "build", "--bin", f"{bin_name}"],
+                cwd=root_dir
+            )
+            shutil.copy2(f"{root_dir}/target/debug/{bin_name}", target_dir)
+            break
+        elif prompt.lower() in ["n", "no"]:
+            logging.critical("Can't launch the network without the binary. Aborting...")
+            sys.exit(4)
+        else:
+            logging.error("Please answer with either `y[es]` or `n[o])")
+
+def main(args):
+    # Bold ASCII escape sequence
+    logging.basicConfig(level=logging.INFO if args.verbose else logging.WARNING,
+        style="{", 
+        format="{asctime} {levelname} \033[1m{funcName}:{lineno}\033[0m: {message}",)
+    # ISO 8601 timestamps without timezone
+    logging.Formatter.formatTime = (lambda self, record, datefmt=None:
+        datetime.datetime.fromtimestamp(record.created)
+        .isoformat(sep="T",timespec="microseconds"))
+    # Colored log levels
+    logging.addLevelName(logging.INFO, f"\033[32m{logging.getLevelName(logging.INFO)}\033[0m")
+    logging.addLevelName(logging.ERROR, f"\033[35m{logging.getLevelName(logging.ERROR)}\033[0m")
+    logging.addLevelName(logging.CRITICAL, f"\033[31m{logging.getLevelName(logging.CRITICAL)}\033[0m")
+    if args.command == "setup":
+        setup(args)
+    elif args.command == "cleanup":
+        cleanup(args)
+
+def setup(args):
+    logging.info(f"Starting iroha network with {args.n_peers} peers...")
+    try:
+        os.mkdir(args.out_dir)
+    except FileExistsError:
+        logging.error(f"Test directory `{args.out_dir}` already exists")
+        sys.exit(5)
+    try:
```

Comment:
Maybe it would be cleaner if all the "try to copy, prompt to build" logic was handled in one function? Otherwise there's a fair amount of code duplication..

Reply:
I did put that logic into a `prompt_build_bin` func though. And other exception handling, while a bit cumbersome, still feels at least as terse as the previous bash script, and definitely more readable.

Reply:
It's just that the

```python
try:
    shutil.copy2(f"{args.root_dir}/target/debug/somebin", args.out_dir)
except FileNotFoundError:
    prompt_build_bin("somebin", args.root_dir, args.out_dir)
```

snippet is repeated a bunch of times, while it can itself be part of the `prompt_build_bin` function (probably would need another name). The repetition is not _that_ bad tbh, but I don't see where it would hurt to reduce it

Reply:
Good point, sorry for misunderstanding.